### PR TITLE
Refactor evasion rects for selection logic

### PR DESCRIPTION
**Pull request by rr-codes**
#### 76a6ea1366a0d7718b3fc36f2bebf3544f1653f3
<pre>
Refactor evasion rects for selection logic
<a href="https://bugs.webkit.org/show_bug.cgi?id=250111">https://bugs.webkit.org/show_bug.cgi?id=250111</a>
rdar://100917527

Reviewed by Aditya Keerthi.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::updateAppearance):
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::getEvasionRectsAroundSelection):
* Source/WebCore/page/EditorClient.h:
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::repositionCaretDecoration):
* Source/WebCore/page/Page.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCaretDecorationPlacement):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::setCaretDecorationPlacement):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setCaretDecorationPlacement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClientCocoa.mm:
(WebKit::WebChromeClient::getEvasionRectsAroundSelection):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
(WebKit::WebEditorClient::setCaretDecorationPlacement):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::clickableNodeAtLocation):
(WebKit::WebPage::isTransparentOrFullyClipped const):
(WebKit::WebPage::getEvasionRectsAroundSelection):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestEvasionRectsAboveSelection):
(WebKit::WebPage::isTransparentOrFullyClipped const): Deleted.
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.h:

Canonical link: <a href="https://commits.webkit.org/258589@main">https://commits.webkit.org/258589@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2240bf8afb87ea45c54b5852b2aa673a56d08c78

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102393 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11527 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35451 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111689 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106366 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12522 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2444 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/109398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108174 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24337 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5021 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25760 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5169 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2201 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11196 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45253 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5895 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6913 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm [hunk 1]
```diff
@@ -508,6 +512,16 @@
     }
 
     postLayoutData.baseWritingDirection = frame.editor().baseWritingDirectionForSelectionStart();
+
+#if PLATFORM(MAC)
+    if (auto* document = frame.document()) {
+        if (auto* page = document->page()) {
+            const Vector<FloatPoint> offsetsForHitTesting { { -30, 50 }, { 30, 50 }, { -60, 35 }, { 60, 35 }, { 0, 20 } };
+            const auto rects = page->chrome().client().getEvasionRectsAroundSelection(offsetsForHitTesting, false);
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm]: `getEvasionRectsAroundSelection` is on `WebPage`, no need for the indirection.

File: Source/WebKit/UIProcess/mac/WebViewImpl.mm [hunk 2]
```diff
@@ -2693,8 +2693,10 @@ static String commandNameForSelector(SEL selector)
         m_softSpaceRange = NSMakeRange(NSNotFound, 0);
 #if HAVE(TOUCH_BAR)
     updateTouchBar();
-    if (m_page->editorState().hasPostLayoutData())
+    if (m_page->editorState().hasPostLayoutData()) {
         requestCandidatesForSelectionIfNeeded();
+        updateCaretDecorationPlacement();
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/mac/WebViewImpl.mm]: Seems strange that `updateCaretDecorationPlacement` is underneath `HAVE(TOUCH_BAR)`.

Comment (rr-codes) [reply Source/WebKit/UIProcess/mac/WebViewImpl.mm]: Very strange indeed 😛 I shall fix

File: Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm (hunk 1 above)
Comment (rr-codes) [reply Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm]: @pxlcoder yes, but `page` is a `Page`, not a `WebPage`

Comment (pxlcoder) [reply Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm]: Yes, but there's no need to go through `page`, this code is already in `WebPage`.

Comment (rr-codes) [reply Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm]: ohhh true, forgot I was in `WebPageCocoa` haha

File: Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm [hunk 3]
```diff
@@ -1964,103 +1951,9 @@ static bool insideImageOverlay(const VisiblePosition& position)
...
-        if (bounds.contains(enclosingIntRect(selectionBoundsInRootViewCoordinates)))
-            continue;
-
-        rectsToAvoidInRootViewCoordinates.append(WTFMove(bounds));
-    }
-
-    reply(WTFMove(rectsToAvoidInRootViewCoordinates));
+    const Vector<FloatPoint> offsetsForHitTesting { { -30, -50 }, { 30, -50 }, { -60, -35 }, { 60, -35 }, { 0, -20 } };
+    auto rects = getEvasionRectsAroundSelection(offsetsForHitTesting);
+    return reply(WTFMove(rects));
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm]: No need for `return`.

File: Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm [hunk 4]
```diff
@@ -508,6 +512,12 @@
     }
 
     postLayoutData.baseWritingDirection = frame.editor().baseWritingDirectionForSelectionStart();
+
+#if PLATFORM(MAC)
+    const Vector<FloatPoint> offsetsForHitTesting { { -30, 50 }, { 30, 50 }, { -60, 35 }, { 60, 35 }, { 0, 20 } };
+    const auto rects = getEvasionRectsAroundSelection(offsetsForHitTesting, false);
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm]: No need for a local variable:

`postLayoutData.evasionRectsAroundSelection = getEvasionRectsAroundSelection(offsetsForHitTesting, false);`

File: Source/WebCore/page/Frame.h [hunk 5]
```diff
@@ -327,11 +327,11 @@ class Frame final : public AbstractFrame {
 #if PLATFORM(IOS_FAMILY)
     void betterApproximateNode(const IntPoint& testPoint, const NodeQualifier&, Node*& best, Node* failedNode, IntPoint& bestPoint, IntRect& bestRect, const IntRect& testRect);
     bool hitTestResultAtViewportLocation(const FloatPoint& viewportLocation, HitTestResult&, IntPoint& center);
-    
+
     enum class ShouldApproximate : bool { No, Yes };
     enum class ShouldFindRootEditableElement : bool { No, Yes };
     Node* qualifyingNodeAtViewportLocation(const FloatPoint& viewportLocation, FloatPoint& adjustedViewportLocation, const NodeQualifier&, ShouldApproximate, ShouldFindRootEditableElement = ShouldFindRootEditableElement::Yes);
-
+
```
Comment (pxlcoder) [inline Source/WebCore/page/Frame.h]: Nit: Unnecessary whitespace.

File: Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm [hunk 6]
```diff
@@ -36,6 +36,8 @@
 #import "WebPasteboardOverrides.h"
 #import "WebPaymentCoordinator.h"
 #import "WebRemoteObjectRegistry.h"
+#import <WebCore/Chrome.h>
+#import <WebCore/ChromeClient.h>
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm]: Are these imports necessary?

File: Source/WebKit/WebProcess/WebCoreSupport/WebChromeClientCocoa.mm [hunk 7]
```diff
@@ -29,6 +29,7 @@
 #if PLATFORM(COCOA)
 
 #import "WebIconUtilities.h"
+#import "WebPage.h"
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebCoreSupport/WebChromeClientCocoa.mm]: Is this import necessary?

File: Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm (hunk 3 above)
Comment (rr-codes) [reply Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm]: I'm honestly surprised this even compiled

File: Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm [hunk 8]
```diff
@@ -668,6 +675,135 @@
...
+    return Ref(m_page->mainFrame())->nodeRespondingToClickEvents(viewportLocation, adjustedViewportLocation);
+#else
+    UNUSED_PARAM(adjustedViewportLocation);
+
+    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frameView = frame->view();
+    if (!frameView)
+        return nullptr;
+
+    auto center = frameView->windowToContents(roundedIntPoint(viewportLocation));
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm]: `pointInContentsCoordinates`? I don't think the point is not necessarily the center of something.

File: Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm [hunk 9]
```diff
@@ -668,6 +675,135 @@
...
+
+    WebCore::HitTestResult result { center };
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm]: You can put this near line 697 – it will be unused if we early return.

**Review by pxlcoder — APPROVED** [review]
